makeflow/src/makeflow_module_archive.c

```diff
@@ -0,0 +1,503 @@
+		debug(D_ERROR|D_MAKEFLOW_HOOK, "could not archive output file %s at %s\n",f->outer_name, file_archive_path);
+		rv = 1;
+		goto FAIL;
+	}
+
+	int symlink_failure = symlink(file_archive_path, job_file_archive_path);
+	if (symlink_failure && errno != EEXIST) {
+		debug(D_ERROR|D_MAKEFLOW_HOOK, "could not create symlink %s pointing to %s\n", job_file_archive_path, file_archive_path);
+		rv = 1;
+		goto FAIL;
+	}
+
+FAIL:
+	free(id);
+	free(file_archive_dir);
+	free(file_archive_path);
+	return rv;
+}
+
+static int makeflow_archive_write_input_files(struct archive_instance *a, struct batch_task *t, char *archive_directory_path) {
+	struct batch_file *f;
+
+	struct list_cursor *cur = list_cursor_create(t->input_files);
+	for(list_seek(cur, 0); list_get(cur, (void**)&f); list_next(cur)) {
+		char *input_file_path = string_format("%s/input_files/%s", archive_directory_path,  f->outer_name);
+		int failed_checksum = write_file_checksum(a, f, input_file_path);
+		free(input_file_path);
+		if(failed_checksum){
+			list_cursor_destroy(cur);
+			return 0;
+		}
+	}
+	list_cursor_destroy(cur);
+	return 1;
+}
+
+static int makeflow_archive_write_output_files(struct archive_instance *a, struct batch_task *t, char *archive_directory_path) {
+	struct batch_file *f;
+
+	struct list_cursor *cur = list_cursor_create(t->output_files);
+	for(list_seek(cur, 0); list_get(cur, (void**)&f); list_next(cur)) {
+		char *output_file_path = string_format("%s/output_files/%s", archive_directory_path,  f->outer_name);
+		int failed_checksum = write_file_checksum(a, f, output_file_path);
+		free(output_file_path);
+		if(failed_checksum){
+			list_cursor_destroy(cur);
+			return 0;
+		}
+	}
+	list_cursor_destroy(cur);
+	return 1;
+}
+
+static int makeflow_archive_create_dir(char * prefix, char * name){
+	char *tmp_directory_path = string_format("%s%s", prefix, name);
+	int created = create_dir(tmp_directory_path, 0777);
+	free(tmp_directory_path);
+	if (!created){
+		debug(D_ERROR|D_MAKEFLOW_HOOK,"Could not create archiving directory %s\n", tmp_directory_path);
+		return 1;
+	}
+	return 0;
+}
+
+static int makeflow_archive_populate(struct archive_instance *a, struct dag_node *n, struct batch_task *t) {
+	char *id = generate_task_archive_id(a, t);
+	int result = 1;
+
+	/* The archive name is binned by the first 2 characters of the id for compactness */
+	char *archive_directory_path = string_format("%s/tasks/%.2s/%s", a->dir, id, id);
+	debug(D_MAKEFLOW_HOOK, "archiving task %d to %s", t->taskid, archive_directory_path);
+
+	int dir_create = 0;
+	/* We create all the sub directories upfront for convenience */
+	dir_create = makeflow_archive_create_dir(archive_directory_path, "/output_files/");
+	dir_create += makeflow_archive_create_dir(archive_directory_path, "/input_files/");
+	dir_create += makeflow_archive_create_dir(archive_directory_path, "/descendants/");
+	dir_create += makeflow_archive_create_dir(archive_directory_path, "/ancestors/");
+
+	if(dir_create){
```

Comment:
This is just being pedantic, but I don't think this reads well. It's clever, but I think it needs a better name. Otherwise, it reads as "if dir_created" when it should be "if error_dir_created"